nft_metadata/program/src/instruction.rs

```diff
@@ -0,0 +1,375 @@
+use std::{mem::size_of, str::FromStr};
+
+use solana_program::{
+    instruction::{AccountMeta, Instruction},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvar,
+};
+
+use crate::{
+    error::NFTMetadataError,
+    state::nft_metadata::{
+        CATEGORY_LENGTH, CREATOR_LENGTH, NAME_LENGTH, SYMBOL_LENGTH, URI_LENGTH,
+    },
+};
+
+/// Used for telling caller what type of format you want back
+#[derive(Clone, PartialEq)]
+pub enum Format {
+    /// JSON format
+    JSON,
+    /// MsgPack format
+    MsgPack,
+}
+impl Default for Format {
+    fn default() -> Self {
+        Format::JSON
+    }
+}
+
+/// Instructions supported by the NFTMetadata program.
+#[derive(Clone)]
+pub enum NFTMetadataInstruction {
+    /// Create an NFT Owner and NFT Metadata objects.
+    ///   0. `[writable]` NFT Owner key (pda of ['metadata', program id, name, symbol])
+    ///   1. `[writable]` NFT metadata key (pda of ['metadata', program id, mint id])
+    ///   2. `[]` Mint of NFT
+    ///   3. `[signer]` Mint authority
+    ///   4. `[signer]` payer
+    ///   5. `[]` NFT metadata program
+    ///   6. `[]` System program
+    CreateNFTMetadataAccounts {
```

Comment:
I think combining create & init into one instruction will make for simpler calls

Reply:
You can't combine it, it won't work. The accounts aren't created, and so the data isn't there to be deserialized, yet.